### PR TITLE
Implement rollup & .d.ts generator

### LINKED ISSUE #1371: Improve Bundling (switch to rollup)
I was sitting and looking over the NPM downloads of this library, and the size of the bundle jumped out at me.

image
I mean, 1.15 MB with 400 files for a library? Seems a little too much to me. 

That's why I thought it would be cool to switch to rollup, like the other libraries.

Results: ~194kb and only 3 files, no more!
image

Also, to be able to merge all the `.d.ts` into single file, I used [@microsoft/api-extractor](https://www.npmjs.com/package/@microsoft/api-extractor), which imo is really good and efficient.

TODO:
- [ ] Fix tests (or switch to JEST)
- [ ] Improve CI

### PR DESCRIPTION
**Pull request by leonardssh**
## Description

This PR implements rollup bundling and `.d.ts` generator using [@microsoft/api-extractor](https://www.npmjs.com/package/@microsoft/api-extractor).

## Related Issue

#1371 

## Types of changes

- [ ] Updated docs / Refactor code / Added a tests case (non-breaking change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have updated the changelog.

### CODE REVIEW

**Comment by leonardssh** [comment]
Apparently, it's not as easy as I'd hoped, there's a bit more work to do on the testing side. 😅

File: tsconfig.json [hunk 1]
```diff
@@ -1,27 +1,21 @@
...
-        "strictNullChecks": true
-    }
+  "compilerOptions": {
+    "target": "esnext",
+    "module": "esnext",
+    "lib": ["es6", "dom"],
+    "types": ["mocha", "reflect-metadata"],
+    "moduleResolution": "node",
+    "sourceMap": true,
+    "jsx": "react",
```
Comment (PodaruDragos) [inline tsconfig.json:9]: do we need this ?

File: scripts/clean-dist.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+(async () => {
```
Comment (PodaruDragos) [inline scripts/clean-dist.js:1]: `clean-dist` and `remove-dist` could be done using `rm -rf dist/` i assume
no need to import packages for that.

File: rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    external,
+    plugins: [tsPlugin, ...plugins],
+    output,
+    onwarn: (msg, warn) => {
+      if (!/Circular/.test(msg)) {
+        warn(msg);
+      }
+    },
+    treeshake: {
+      moduleSideEffects: false
```
Comment (PodaruDragos) [inline rollup.config.js:60]: should this not be `true` ?

File: package.json [hunk 4]
```diff
@@ -69,11 +78,37 @@
...
+    "rollup": "^2.56.3",
+    "rollup-plugin-typescript2": "^0.30.0",
     "sinon": "9.2.4",
     "ts-node": "9.1.1",
     "tsify": "5.0.2",
     "tslint": "6.1.3",
     "typescript": "4.2.4",
     "updates": "12.0.2"
+  },
+  "prettier": {
```
Comment (PodaruDragos) [inline package.json:90]: what's up with prettier here ?

**Comment by notaphplover** [comment]
Closing the PR. It's too old and we're no longer going this way

**Pull request closed without merging**